### ISSUE TITLE
Skeleton add the option to Lock Bone Scaling

### ISSUE THREAD

**Issue by mrjustaguy** [godotengine/godot-proposals#4190]
### Describe the project you are working on

Character Creator

### Describe the problem or limitation you are having in your project

Scaling a given bone will also apply it's scaling to it's Direct children bones (indirectly also the bones of those child bones) and the Direct children Bones have to get the exact opposite scaling to undo this (for example scale a parent bone by 2, the children of the parent bone have to be scaled by 0.5 for them to keep their scale)

This behavior is counter-intuitive, and while it makes sense when you think about the inner workings, it doesn't provide the most common use case in it's area which most users that do this will use and expect.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Allowing Scale to not be inherited from parent bones would allow for easy individual bone editing often used in Character Creators, both in apps and games (most often RPGs)

Locking Translation/Rotation like this afaik doesn't make any sense for anything.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Add a bool to bones to ignore parent bone scaling, and use by default, as this is generally the desired outcome (I haven't seen many use the current behavior where parent bones also scale it's child bones, albeit there are some applications of this too so keeping the option available makes sense)

This part is self explanatory, if a Bone is ignoring it's parent bone scaling, it doesn't scale itself (and it's children too as a result) when the parent bone changes it's scaling. This is just a tiny bit of simple math.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It can be.

### Is there a reason why this should be core and not an add-on in the asset library?

This is an intuition issue in Core, and while it can be fixed with an addon, one shouldn't have to use addons to fix the editor's lack of intuitiveness.

**Comment by Calinou** [comment]
cc @TokageItLab @fire

**Comment by TokageItLab** [comment]
That behavior certainly makes sense for character creation, but not for animation.

Also, when doing character creation, it is common to apply the final scale to the rest, not the pose. Think of Blender's EditBone edit; At least, I think that rests should be operated independently of the parent-child relationship.

In my opinion, the proper way to go is to implement a RestEdit mode that allows such rests to be edited independently, in addition to the PoseEdit mode that currently already exists in SkeletonEditor.

**Comment by mrjustaguy** [comment]
Probably, however I don't see any use of setting translation/rotation of bones independently.. but yeah, Don't see why it wouldn't be sufficient.

It would also indeed be very helpful for editor UX in general to have the ability to specifically edit rest, regardless of the individual bone setting asked for in this proposal, and then have it set bones independently, while keeping the pose editor the same.

**Comment by raldone01** [comment]
I created a model in blender and set ignore parent scale on some child bones.
The gltf export preserves these attributes.
However godot ignored them which is counter intuitive.
I will try to write a script to rescale the affected child bones.

**Comment by ElfWitch** [comment]
Is this something that should be implemented as a single bool on a Skeleton3D that affects all bones, or should it be per bone?

**Comment by mrjustaguy** [comment]
It's a really good question..

One part of me says default to on as it's more intuitive, but the other doesn't want to screw over existing animations..

I think keeping it off is probably the way to go, it's not like scaling is a super common thing in animation and just having the option between the two behaviors should help significantly for those using it, so doesn't really matter a ton if the default option is the less intuitive one, and I'd say compatibility takes priority here..

**Comment by ElfWitch** [comment]
I agree about compatibility. I'm pretty sure it's done. This is my first time programming c++ and my first PR ever so hopefully I did everything right. 🤞

https://github.com/godotengine/godot/commit/c320ed4411db597c58d90e169221624e363d3039

**Comment by TokageItLab** [comment]
There was a previous proposal almost identical to this one. As I added there, this needs to be an EDITOR ONLY feature.

The use of the Bone Scale in the Character Creator assumes a Rest variant. At this point, it makes sense to make it an option in the editor whether or not to propagate the parent's deformations. It would work like Blender's bone Edit Mode.

The problem at this time is that Pose in Godot 4.x is no longer relative to Rest.

I've argued this issue forcefully in https://github.com/godotengine/godot/pull/56902 and published a module in https://github.com/TokageItLab/realtime_retarget, but maybe we should re-arrange to bring that implement it into core once again now that the demand has been visualized.

**Comment by ElfWitch** [comment]
> this should be an EDITOR ONLY feature.
>
>The use of the Bone Scale in the Character Creator assumes a Rest transform. At this point, it makes sense to make it an option in the editor whether or not to propagate the parent's transforms. It would work like Blender's bone Edit Mode.

If this was editor only, it would make certain runtime character proportion editing and procedural animation impossible. By making it an option that can be applied per bone at runtime, we open the possibilities for non-authored procedural animations that weren't possible before. I'd also like to mention that even though blender does use rest transforms for its bone Edit Mode, even it has an option to control how transforms are propagated during animation per bone:
![image-1](https://github.com/godotengine/godot-proposals/assets/109904725/f01db023-39a1-4e22-8286-ba7d40aca379)
This PR introduces modes that would be equivalent to Blender's functionality as far as:
`inherit_scale=true` = Blender's Full
`inherit_scale=false` = Blender's None -which uses similar sheer correcting orthonormalization under the hood.
These are the modes used most often in blender animations, and blender is the animation authoring tool of choice for most Godot users, so it makes sense to support a one-to-one feature like this, regardless of potential future work with rest transforms.

> "Pose Scale is propagated to the children" is never a problem.

This *can* be a problem if you are trying to make character customization that relies on proportional editing, which is the primary use case for needing to change bone scale inheritance. https://github.com/godotengine/godot/pull/83903 addresses two problems. Firstly, it adds a feature that allows users to choose which bones inherit their parent's full transform, or only inherit rotation/translation. In the current version of Godot, you can manually change a child bone's transform to account for the scale propagation, but the problem is that even with manual scale correction, there are major sheering and rotation artifacts which get worse the further the child scale is from 1.0. The second problem this PR addresses is the aforementioned sheering and rotation artifacts, when using `inherit_scale=false`.

**Comment by fire** [comment]
> Inverse Kinematics is a method of animating objects using bones chained into a skeleton. Instead of manually setting the rotation of individual bones in the skeleton, IK solves the rotations for us based on the end target position.

I thought that to lock a bone's position or rotation, you would typically use an Inverse Kinematics (IK) constraint system.
